src/posix/platform/infra_if.cpp

```diff
@@ -0,0 +1,362 @@
+
+    uint8_t             cmsgBuffer[CMSG_SPACE(sizeof(*packetInfo))];
+    struct msghdr       msgHeader;
+    struct cmsghdr *    cmsgPointer;
+    ssize_t             rval;
+    struct sockaddr_in6 dest;
+
+    VerifyOrExit(sInfraIfIcmp6Socket >= 0, error = OT_ERROR_FAILED);
+    VerifyOrExit(aInfraIfIndex == sInfraIfIndex, error = OT_ERROR_DROP);
+
+    memset(cmsgBuffer, 0, sizeof(cmsgBuffer));
+
+    // Send the message
+    memset(&dest, 0, sizeof(dest));
+    dest.sin6_family = AF_INET6;
+    memcpy(&dest.sin6_addr, aDestAddress, sizeof(*aDestAddress));
+#ifdef HAVE_SIN6_SCOPE_ID
+    if (IN6_IS_ADDR_LINKLOCAL(&addr.sin6_addr) || IN6_IS_ADDR_MC_LINKLOCAL(&addr.sin6_addr))
+    {
+        dest.sin6_scope_id = sInfraIfIndex;
+    }
+#endif
+
+    iov.iov_base = const_cast<uint8_t *>(aBuffer);
+    iov.iov_len  = aBufferLength;
+
+    msgHeader.msg_namelen    = sizeof(dest);
+    msgHeader.msg_name       = &dest;
+    msgHeader.msg_iov        = &iov;
+    msgHeader.msg_iovlen     = 1;
+    msgHeader.msg_control    = cmsgBuffer;
+    msgHeader.msg_controllen = sizeof(cmsgBuffer);
+
+    // Specify the interface
+    cmsgPointer             = CMSG_FIRSTHDR(&msgHeader);
+    cmsgPointer->cmsg_level = IPPROTO_IPV6;
+    cmsgPointer->cmsg_type  = IPV6_PKTINFO;
+    cmsgPointer->cmsg_len   = CMSG_LEN(sizeof(*packetInfo));
+    packetInfo              = (struct in6_pktinfo *)CMSG_DATA(cmsgPointer);
+    memset(packetInfo, 0, sizeof(*packetInfo));
+    packetInfo->ipi6_ifindex = sInfraIfIndex;
+
+    rval = sendmsg(sInfraIfIcmp6Socket, &msgHeader, 0);
+    if (rval < 0)
+    {
+        otLogWarnPlat("failed to send ICMPv6 message: %s", strerror(errno));
+        ExitNow(error = OT_ERROR_FAILED);
+    }
+
+    if (static_cast<size_t>(rval) != iov.iov_len)
+    {
+        otLogWarnPlat("failed to send ICMPv6 message: partially sent");
+        ExitNow(error = OT_ERROR_FAILED);
+    }
+
+exit:
+    return error;
+}
+
+static void InitLinkLocalAddress()
+{
+    struct ifaddrs *ifAddrs;
+
+    if (getifaddrs(&ifAddrs) < 0)
+    {
+        otLogCritPlat("failed to get netif addresses: %s", strerror(errno));
+        DieNow(OT_EXIT_ERROR_ERRNO);
+    }
+
+    for (struct ifaddrs *addr = ifAddrs; addr != nullptr; addr = addr->ifa_next)
+    {
+        struct sockaddr_in6 *ip6Addr;
+
+        if (strncmp(addr->ifa_name, sInfraIfName, sizeof(sInfraIfName)) != 0 || addr->ifa_addr->sa_family != AF_INET6)
+        {
+            continue;
+        }
+
+        ip6Addr = reinterpret_cast<sockaddr_in6 *>(addr->ifa_addr);
+        if (IN6_IS_ADDR_LINKLOCAL(&ip6Addr->sin6_addr))
+        {
+            memcpy(&sInfraIfLinkLocalAddr, &ip6Addr->sin6_addr, sizeof(sInfraIfLinkLocalAddr));
+            break;
+        }
+    }
+}
+
+void platformInfraIfInit(otInstance *aInstance, const char *aIfName)
+{
+    int                 sock;
+    struct icmp6_filter filter;
+    ssize_t             rval;
+    static const int    kOne      = 1;
```

Comment:
```suggestion
    const int    kOne      = 1;
```
what's the benefit of static?

Reply:
With `static`: `kOne` is initialized only once.
Without `static`: `kOne` is initialized every times this method is called. Because we are taking the address of `&kOne` later, so `kOne` will never be optimized as a constant.

Discard the fact that this method is called only once, I prefer using `static const` wherever possible.

Reply:
`static` adds to the total memory footprint of the firmware.

Reply:
That makes sense.